pkg/clients/handlers.go

```diff
@@ -0,0 +1,166 @@
+package clients
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github.com/canonical/identity-platform-admin-ui/internal/logging"
+	"github.com/canonical/identity-platform-admin-ui/internal/responses"
+	"github.com/go-chi/chi/v5"
+	hClient "github.com/ory/hydra-client-go/v2"
+)
+
+type API struct {
+	service ServiceInterface
+
+	logger logging.LoggerInterface
+}
+
+func (a *API) RegisterEndpoints(mux *chi.Mux) {
+	mux.Get("/clients", a.ListClients)
+	mux.Post("/clients", a.CreateClient)
+	mux.Get("/clients/{client_id}", a.GetClient)
+	mux.Put("/clients/{client_id}", a.UpdateClient)
+}
+
+func (a *API) WriteJSONResponse(w http.ResponseWriter, data interface{}, msg string, status int, links interface{}, meta interface{}) {
+	w.Header().Set("Content-Type", "application/json")
+	r := new(responses.Response)
+
+	r.Data = data
+	r.Message = msg
+	r.Status = status
+	r.Links = links
+	r.Meta = meta
+
+	rr, err := json.Marshal(r)
+	if err != nil {
+		a.logger.Errorf("Unexpected error: %s", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Write(rr)
+	w.WriteHeader(r.Status)
+}
+
+func (a *API) GetClient(w http.ResponseWriter, r *http.Request) {
+	clientId := chi.URLParam(r, "client_id")
+
+	c, errResp, e := a.service.GetClient(r.Context(), clientId)
+	if errResp != nil {
+		a.WriteJSONResponse(w, errResp, "", int(*errResp.StatusCode), nil, nil)
+		return
+	}
+	if e != nil {
+		a.logger.Errorf("Unexpected error: %s", e)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	a.WriteJSONResponse(w, c, "", http.StatusOK, nil, nil)
+}
+
+func (a *API) CreateClient(w http.ResponseWriter, r *http.Request) {
+	c := hClient.NewOAuth2Client()
+
+	// TODO @nsklikas: Limit request params?
+	json_data, err := io.ReadAll(r.Body)
+	if err != nil {
+		a.logger.Debugf("Failed to read response body: %s", err)
+		w.WriteHeader(http.StatusBadRequest)
+	}
+	err = json.Unmarshal(json_data, c)
+	if err != nil {
+		a.logger.Debugf("Failed to unmarshal JSON: %s", err)
+		w.WriteHeader(http.StatusBadRequest)
+	}
+
+	c, errResp, e := a.service.CreateClient(r.Context(), c)
+	if errResp != nil {
+		a.WriteJSONResponse(w, errResp, "", int(*errResp.StatusCode), nil, nil)
+		return
+	}
+	if e != nil {
+		a.logger.Errorf("Unexpected error: %s", e)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
```

Comment:
as discussed, there might be a way to offload all of this to the service layer, simply passing the context and the unmarshalled/deserialezed body 

see if there is a nicer way, if not this does the job